### PR TITLE
Resolves JS-278

### PR DESCRIPTION
**Pull request by ericmorand-sonarsource**
Fixes JS-278

Please read the commit messages for more information.

Also, this may seem like a big change according to the number of files changed, but around 440 of these files are Java Check subclasses where the `eslintKey` public getter was removed. You can safely ignore them and focus on the 20 or so relevant changes that this PR brings.

### CODE REVIEW

File: packages/jsts/tests/linter/wrapper.test.ts [hunk 1]
```diff
@@ -352,7 +352,9 @@ describe('LinterWrapper', () => {
       }
 
       if (!fixture) {
-        throw new Error(`Failed to find fixture file for rule '${ruleId}' in '${fixtures}'.`);
+        // todo: discuss with the team if we can not throw on missing fixture
```
Comment (vdiez) [inline packages/jsts/tests/linter/wrapper.test.ts:355]: we didn't discuss this. can we have it thowing again? what happens if we do?

File: packages/jsts/tests/rules/index.test.ts [hunk 2]
```diff
@@ -52,7 +52,8 @@ describe('index', () => {
       }
     }
     expect(missing).toHaveLength(0);
-    expect(Object.keys(rules)).toHaveLength(mappedRules.size);
+    // todo: this test is fragile and does not guarantee anything relevant: that a certain number of symbols are exported does not mean that the expected symbols are exported; we need to implement a relevant test
+    //expect(Object.keys(rules)).toHaveLength(mappedRules.size);
```
Comment (vdiez) [inline packages/jsts/tests/rules/index.test.ts:56]: agreed, feel free to remove completely then

File: sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/Check.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
+package org.sonar.plugins.javascript.api;
+
+import org.sonar.check.Rule;
+
+public class Check implements EslintBasedCheck {
```
Comment (vdiez) [inline sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/Check.java:5]: I wonder if we can get rid completely of EslintBasedCheck, instead of hidding it with an intermediate class

**Review by vdiez — APPROVED** [review]
LGTM, very nice Eric! thx!

some small comments, from what we discussed you can focus on the quickfixes test and ignore the others.

File: packages/jsts/tests/linter/wrapper.test.ts (hunk 1 above)
Comment (ericmorand-sonarsource) [reply packages/jsts/tests/linter/wrapper.test.ts:355]: If we do, we must add fixtures named SXXX alongside the ESLint-named ones. Definitely feasible, but time-consuming - we have to copy and paste under another name each available fixture. This added value in quality is debatable in my opinion: if the ESLint-named fixture is successful, then the SXXX alias is covered too.

File: sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/TestFileCheck.java [hunk 4]
```diff
@@ -22,8 +22,10 @@
 import java.util.Collections;
 import java.util.List;
 import org.sonar.api.batch.fs.InputFile;
+import org.sonar.check.Rule;
 
-public abstract class TestFileCheck implements EslintBasedCheck {
+@Rule(key="key")
```
Comment (saberduck) [inline sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/TestFileCheck.java:27]: why is this needed? imo it shouldn't be needed on abstract class

File: sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/Check.java (hunk 3 above)
Comment (ericmorand-sonarsource) [reply sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/Check.java:5]: For posterity, we discussed this by private message and we agree that the ESLintbasedCheck interface can, and should, be removed. There would be some impact on plugins that consume the API of SonarJS, like SonarSecurity, which is why we keep it for now.

File: sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/TestFileCheck.java (hunk 4 above)
Comment (ericmorand-sonarsource) [reply sonar-plugin/api/src/main/java/org/sonar/plugins/javascript/api/TestFileCheck.java:27]: True, I think it is a mistake. Let me remove it and see how it goes.

File: packages/jsts/tests/rules/index.test.ts (hunk 2 above)
Comment (ericmorand-sonarsource) [reply packages/jsts/tests/rules/index.test.ts:56]: Thanks. At one point, we may want to add a test that checks explicitly what we export as a public API for the ESLint Plugin.

File: packages/jsts/tests/linter/wrapper.test.ts (hunk 1 above)
Comment (vdiez) [reply packages/jsts/tests/linter/wrapper.test.ts:355]: but you have the mapping between both keys, don't you? I would modify the test to look for the right file no matter the kind of id. if we leave it like this, some valid ids might be ignored. if you want to avoid repetition, just ignore ids `/^S\d+$/`. WDYT?

Comment (ericmorand-sonarsource) [reply packages/jsts/tests/linter/wrapper.test.ts:355]: > but you have the mapping between both keys, don't you? 

Not explicitly - i.e. not in an operable manner. 

>  if you want to avoid repetition, just ignore ids /^S\d+$/.

Yeah, sounds good to me. Will do.
